### PR TITLE
Shared context introduction section

### PR DESCRIPTION
**Pull request by emilyriederer**
This PR is a first draft of the 'shared context' session we discussed via email. This is not yet ready 'for production', but I would appreciate any feedback on the scope of the content included, writing style, or any other direction you wish to provide. I'm happy to make any sort of changes and explore other approaches.

Here are a few questions that I debated while writing, and I would appreciate any advice you have:

- Is this the right content? 
  + I think my slow realization of all of these topics has definitely helped me navigate RMarkdown but I obviously don't want to overfit to my own experience.
- Is the level of explanation appropriate? 
  + Since our goal is to establish a shared context, I tend towards overexplaining instead of underexplaining and, to some extent, explaining helpful (I hope!) heuristics versus the full technical mechanisms. This ultimately targets the less experienced readers. 
  + On one hand, this hopefully helps level the playing field for them, but on the other hand, I don't want to give more experienced readers a bad first impression of the book! 
- Is the length too long for the amount of content?
- Am I too heavy on cross-references to sections that readers will not have seen yet? 
  + My hope is to make a 'marketing pitch' and grow excitement for the future chapters by teasing what is to come, but perhaps it is confusing to reference future material.
- Should the writing style be more formal or casual? I also know that my writing tends to be a bit more verbose / conversational. I can make this more direct if we prefer.
- Of course, also open to other feedback or directions!

After we resolve any content issues, below are some notes on technical things that I would need to fix before this PR were ready to merge:

- Currently, the 'What happens when we render?' section is blank. I'm wondering if it would be appropriate to pull 2.1 up into this context piece. (Unless this chapter should be appendix material, at which point I can make a reference to 2.1)
- A few sections I would like to cross-reference do not have named headers. May I add those to the other sections? Alternatively, I can verbally reference concepts without formally cross-referencing.
- I made a few mockup `nomnoml` figures to help illustrate what I was talking about. Please let me know if I should remove these (I get the sense some publishers prefer fewer figures?) or rewrite in a different format (other package and/or replace with static image) to avoid the `nomnoml` dependency. This was just an easy way to quickly illustrate what I was picturing.
- The file needs to be renamed and linked to the book

Best,

Emily

### CODE REVIEW

File: 99-context.Rmd [hunk 1]
```diff
@@ -0,0 +1,176 @@
+# RMarkdown Conceptual Overview {#conceptual-overview}
```
[the remark this answers is not in the document]
Comment (emilyriederer) [reply 99-context.Rmd:1]: Thanks for flagging! Sorry, I should have caught on to the conventions. I appreciate the link and will go through the rest of the doc to make sure I'm consistent.

File: 99-context.Rmd [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+```
+title: My RMarkdown
+author: Yihui Xie
+output: html_document
+params:
+  start_date: '2020-01-01'
+  end_date: '2020-06-01'
+```
+````
+
```
[the remark this answers is not in the document]
Comment (emilyriederer) [reply 99-context.Rmd:52]: Great point -- thank you! I will highlight that more explicitly

File: 99-context.Rmd [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+```
+
+These headers give structure to our entire document as `rmarkdown` converts the `.md` to our final output format. We can leverage this structure in a number of ways by making reference to these various sections and appending certain formatting and choices to them. We can do this by giving them a unique name by following the header title with `{#name}` or by attaching one or more classes (or 'groups') to them with `{.type}`.
+
+For example, we might write:
+
+```
+## Second-level header {#introduction .colorful-text}
+```
+
```
[the remark this answers is not in the document]
Comment (emilyriederer) [reply 99-context.Rmd:76]: Thank you! If you think it's alright to mention twice, I like the idea of adding the link. It definitely continues the theme of showing how all of the pieces relate to one another.

File: 99-context.Rmd [hunk 4]
```diff
@@ -0,0 +1,176 @@
+# RMarkdown Conceptual Overview {#conceptual-overview}
+
+The goal of this text is to showcase many tips and tracks for getting the most value from RMarkdown. The following chapters demonstrate techniques to write more efficient and succinct code and to customize your output. Before we begin, it may be helpful to understand just a bit more about how RMarkdown works to help you to understand, remember, apply, and 'remix' these tricks. In this section, we provide a brief overview of the process of knitting a document and the key "levers to pull" to change the output. This material is not necessary to understand the subsequent chapters (so feel free to skip ahead!), but it may help you to build a richer mental model for how all the pieces fit together.
+
+## What happens when we render? {#when-we-render}
+
+**TODO: Could we put Ch2.1 here? Or, if this is Appendix material, we can add a reference to it**
```
[the remark this answers is not in the document]
Comment (emilyriederer) [reply 99-context.Rmd]: Thanks! I didn't want to be presumptuous moving other parts of content around without asking. If you're good with it, I'll shift it over for consideration. Really love this section!

**Comment by emilyriederer** [comment]
Hi @cderv:

Thank you so much for the feedback! This was very helpful.

First, please accept my apology for the formatting errors with no space in 'R Markdown' and using backticks not bold on packages. Definitely noted for the future to reference the conventions that you have linked! Please always feel free to tell me to go back and fix such things. I really appreciate your doing it for me, but I'm sorry to have taken up your time. 

Thanks also for the permission to move in section 2.1 and to add other section IDs to reference. I hesitated to modify any other chapters without checking, but I've now made these changes to have a more complete first draft. 

A few more thoughts on some of your specific comments:

- You mentioned this chapter being less technical than 2.1. This was definitely a concern of mine and something I considered when writing; I wanted to walk the line between being very friendly and accessible, but I certainly don't want to trivialize the concepts by giving a high-level explanation. Please do not hesitate to let me know if you think I should try to match the precision / detail of 2.1 a bit more!

- I do see your point that I do go into unusual depth about some details like classes and groups. My motivation was that I think many tricks throughout the book rely on applying difference classes or referencing different IDs, so I thought these might be "foundational concepts" for readers to grasp. I also thought there was an interesting parallel between how classes and groups are assigned to sections versus code chunks. However, out of context, I can imagine this is somewhat syntax-laden and strikes a different tone than the rest of the chapter. I am happy to remove if you think that's better!

- Are there standards to avoid dependencies? I can either add `nomnoml` as a dependency or I could convert the resulting diagram to an image instead. Anecdotally, I think I also know that some publisher wish to limit figures. Is there any reason I should be concerned about adding figures at all? 

Finally, please let me know if I can help at all reading through the old reviews and categorizing the feedback. I heard there are quite a few so I'm happy to be a second pair of eyes if it could help.

Thanks again!

File: 99-context.Rmd (hunk 4 above)
[the remark this answers is not in the document]
Comment (cderv) [reply 99-context.Rmd]: Let's wait for yihui thoughts on all this.

**Comment by cderv** [comment]
> First, please accept my apology for the formatting errors with no space in 'R Markdown' and using backticks not bold on packages. 

You don't need to apology ! I discovered the same first time I contributed ☺️ It was easy for me to suggest a fix that would be easy for you to include - So I did it 😉 

Let's wait on Yihui thoughts on all this. I just wanted to give mine before he did. At the end, he will know the answer to all your questions (I don't know about publisher and figure for example) and he will decide how and where it fits better. 

Thank you again!

File: 99-context.Rmd [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+````md
+```
+title: My R Markdown report
+author: Yihui Xie
+output: html_document
+```
+````
+
+In this case, the `title` and `author` fields are processed by Pandoc to set the values of template variables. Using the default template, the title and author information will appears at the beginning of the resulting document. More details on how Pandoc uses information from the header are explained on the [R Markdown website's introduction to Pandoc Markdown](https://rmarkdown.rstudio.com/authoring_pandoc_markdown.html%23raw-tex#Metadata_blocks).
```
Comment (cderv) [inline 99-context.Rmd:51]: Maybe it is better to link to Pandoc Manual directly for this ? 🤔 
https://pandoc.org/MANUAL.html#extension-yaml_metadata_block

Comment (emilyriederer) [reply 99-context.Rmd:51]: Excellent point - thank you! I will switch it over.

**Comment by yihui** [comment]
@emilyriederer As the first step, I fixed the Travis error, but I force-pushed to your repo. You probably need to delete your local repo (unless you have any changes that you haven't committed), and clone it again.

Lesson one: use a new git branch instead of `master` for pull requests, because the branches for pull requests can change dramatically (e.g. the commits could be squashed before merging).

I'll commit the rest of changes one by one from now on.

**Comment by emilyriederer** [comment]
Hi @yihui ! Thank you so much for the corrections and for all of these helpful feedback and comments. I'm creating a checklist from all of your an Chris' comments so I can get a better overall sense of the book style and make sure that I do not make any of the same mistakes multiple times. 

Sorry for your trouble on sending the PR from my fork's master! I will definitely be sure to send these from a different branch in the future.

**Comment by yihui** [comment]
No worries! I'm almost done now. More commits to come in about an hour...

**Comment by yihui** [comment]
I have been "tortured" many times by my copy editor at Chapman & Hall, so please treat me as an extremely picky copy editor in the above changes :) I think you will learn more lessons from her soon after we send them the finalized draft.

Unless any of you two have further comments or objections, I think this PR is ready to be merged. There is one important thing that I haven't figured out yet, though: for the **nomnoml** figures, they wouldn't compile for the PDF version of this book (you can run `make pdf` to get the full PDF). I'm not sure how difficult this task is. In theory, it should be simple to deal with HTML widgets for PDF output after **webshot** and PhantomJS are installed. I have **DiagrammeR** graphs in this book that worked fine for PDF output. Not sure why **nomnoml** wouldn't work.

**Comment by emilyriederer** [comment]
Thanks for all of the feedback, @yihui ! I really appreciate it.

Regarding `nomnoml`, please let me know if it would be more convenient if I create static images instead. `nomnoml` was great for a quick mock-up but there's no real reason these views need to regenerate every time the book is rendered. If a static image is preferable, please just let me know if there are any constraints or preferences (e.g. file type / size / dimensions) from you or the publisher.

**Comment by yihui** [comment]
Well, it might be because it was too late last night. I just tried again, and the charts were generated successfully for the PDF version. I'm going to merge this PR for now. If you have further comments, please feel free to open new pull requests (and I suggest you delete your current fork, and fork this repo again to avoid the potential git pain).

**Pull request merged**